cg/apps/demultiplex/sample_sheet/override_cycles_validator.py

```diff
@@ -0,0 +1,172 @@
+    ) -> bool:
+        """
+        Returns whether an index cycle string is following a valid cycle regex pattern and has
+        consistent values. Valid patterns are 'I(\d+)N(\d+)' and 'N(\d+)I(\d+)'. Having consistent
+        values means that the sum of the number of index characters (I) and the number of ignored
+        characters (N) specified in the index cycle string is equal to the number of run cycles
+        and the length of the index sequence is equal to the number of index characters (I).
+        """
+        match = re.match(pattern, index_cycle)
+        if match:
+            if pattern == FORWARD_INDEX_CYCLE_PATTERN:
+                index_chars, ignored_chars = map(int, match.groups())
+            elif pattern == REVERSE_INDEX_CYCLE_PATTERN:
+                ignored_chars, index_chars = map(int, match.groups())
+            else:
+                LOG.warning(f"Pattern {pattern} is not a valid index cycle pattern")
+                return False
+            if index_chars + ignored_chars == run_cycles and len(index_sequence) == index_chars:
+                return True
+        return False
+
+    def _validate_reads_cycles(self) -> None:
+        """
+        Determine if the sample read cycles are valid, i.e. if the sample read cycle values are
+        equal to the respective run read cycles.
+        Raises:
+            OverrideCyclesError if the reads cycles are not valid.
+        """
+        read1_cycle: str = self.sample_cycles[0]
+        read2_cycle: str = self.sample_cycles[-1]
+        if (
+            read1_cycle == f"Y{self.run_read1_cycles}"
+            and read2_cycle == f"Y{self.run_read2_cycles}"
+        ):
+            return
+        message: str = f"Incorrect read cycles {self.sample_cycles} for sample {self.sample_id}"
+        LOG.error(message)
+        raise OverrideCyclesError(message)
+
+    def _validate_index1_cycles(self) -> None:
+        """
+        Determine if the sample index 1 cycle is valid, i.e., if the number of index characters in
+        the override cycles coincides with the length of the index sequence and if the number of
+        ignored characters in the override cycles matches the difference between the length of the
+        index sequence and the number of run index1 cycles.
+        Raises:
+            OverrideCyclesError if the index 1 cycle is not valid.
+        """
+        index1_cycle: str = self.sample_cycles[1]
+        if (
+            self.run_index1_cycles == len(self.sample[SampleSheetBCLConvertSections.Data.INDEX_1])
+            and index1_cycle == f"I{self.run_index1_cycles}"
+        ):
+            return
+        if self._is_index_cycle_value_following_pattern(
+            pattern=FORWARD_INDEX_CYCLE_PATTERN,
+            index_cycle=index1_cycle,
+            run_cycles=self.run_index1_cycles,
+            index_sequence=self.sample[SampleSheetBCLConvertSections.Data.INDEX_1],
+        ):
+            return
```

Comment:
Does this not go against the type hint? If the first condition is met, the second will never be checked? Currently it would be true if condition 1 or 2 is met, instead of 1 and 2

Reply:
Might be nice to do all the negative checks and no positive checks? I.e. no need for a return statement, but errors are raised if a negative check fails.

Reply:
That is precisely the expected behaviour, if the first condition is fulfilled it passes validation. I thought about putting both of them in the same if statement but would look too crowded. I guess @seallard suggestion of creating smaller functions for the if statements could make this and the next function more clear

Reply:
Oh, then I would only like to see a change in the docstring to make it clear that either condition is valid, and that you don't need them both for a pass.

Reply:
Simplified the docstring to avoid confusion